### DISCUSSION TITLE
How to create parent-child relationship between contacts?

### DISCUSSION THREAD

**Discussion by LaurensCleyman** [YetiForceCompany/YetiForceCRM#14073] [How to]
Hi all, 

**For example:**
John is the father of Mike.
Els is the mother of Mike.
Peter is the stepfather of Mike.
Mike is the son of John and Els and the stepson of Peter.

**Goal:**
A contact can have 1 or more parents.
A contact can have 1 or more children.
A relationship between two contacts should only be created once.

Anybody an idea how this could be done? 

Kind regards,

Laurens

**Comment by rskrzypczak** [comment]
I think this is what you mean ....
![image](https://user-images.githubusercontent.com/10221999/105139328-76695000-5af6-11eb-800c-bf49604fa6e1.png)
YetiForce created similar functionality for one of its clients. Unfortunately this is a dedicated functionality that isn't available for free. 
If you are interested in it please get in touch with YetiForce.

**Reply by LaurensCleyman** [reply]
Hi @rskrzypczak , 

That is exactly what is needed, thanks. I'll first try the solution of Apcloic, if not possible I'll contact Yetiforce team!

Kind regards, 

Laurens

**Reply by krseguros** [reply]
I'm interested and I need this

**Comment by apcloic** [comment]
@LaurensCleyman 
I have configured this kind of relationship for a non-profit organization a few months ago, it works well.
For the type of relationship, I've just used the status picklist with some extra options.

I don't remember exactly how but did you simply try to add an M:M relation in Contact module with Contact module ?
https://gitstable.yetiforce.com/index.php?module=LayoutEditor&parent=Settings&view=Index&mode=showRelatedListLayout

YF dedicated functionnality looks better and more advanced but you may try this first.

Regards,

**Reply by LaurensCleyman** [reply]
Hi @apcloic , 

Thanks for pointing me in that direction. I created a M:M relationship like this:

![image](https://user-images.githubusercontent.com/16014524/105761181-18ec5d80-5f53-11eb-96d9-4d8b05d78a97.png)

![image](https://user-images.githubusercontent.com/16014524/105760775-8946af00-5f52-11eb-8a54-dad7b9ee9f44.png)

However, I'm strugling with the picklist field that defines the relationship. Becasue if I add a picklist field 'Relationship type' to contacts, then i can only select one option when I need multiple:
For example: Mike can be a Son for one contact, but a Stepson for another contact.

Current result:
![image](https://user-images.githubusercontent.com/16014524/105761409-68328e00-5f53-11eb-8f50-9979a2547d58.png)


Kind regards,

Laurens

**Reply by apcloic** [reply]
@LaurensCleyman
Hi,
The needs of my non-profit organization was simplest than yours.
However, you may find what you need in Occurences module.
Check Participants related tab which is related to Contacts module, this is handled by a custom relationship (GetRelatedMembers) and there is 2 extra fields "status" and "comment", I think that you need the [same feature](https://gitstable.yetiforce.com/index.php?module=Occurrences&relatedModule=Contacts&view=Detail&record=3138&mode=showRelatedList&relationId=624&tab_label=LBL_PARTICIPANTS).
So if you have programming skills, you have to study this module and particularly this [file](https://github.com/YetiForceCompany/YetiForceCRM/blob/developer/modules/Occurrences/relations/GetRelatedMembers.php)
I will also give a try myself because I need this kind of feature too.
Regards,

EDIT :

Quick test, it looks quite easy.
1- Set $LOAD_CUSTOM_FILES to TRUE in config/Performance.php
2 - Put the following file in /custom/modules/Contacts/relations/GetRelatedMembers.php

<details>
<summary>
Click to view code
</summary>

````

<?php
/**
 * Includes RelatedMembers relation.
 *
 * @package   Relation
 *
 * @copyright YetiForce Sp. z o.o
 * @license   YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
 * @author    Radosław Skrzypczak <r.skrzypczak@yetiforce.com>
 */
/**
 * Class GetRelatedMembers.
 */
class Contacts_GetRelatedMembers_Relation extends Vtiger_GetRelatedList_Relation
{
	/**
	 * {@inheritdoc}
	 */
	public const TABLE_NAME = 'u_#__relations_members_entity';

	/**
	 * Field custom list.
	 *
	 * @var array
	 */
	public $customFields = [
		'status_rel' => [
			'label' => 'LBL_STATUS_REL',
			'uitype' => 16
		],
		'comment_rel' => [
			'label' => 'LBL_COMMENT_REL',
			'uitype' => 21]
	];

	/**
	 * Field list.
	 *
	 * @return array
	 */
	public function getFields()
	{
		$fields = [];
		$sourceModule = $this->relationModel->getParentModuleModel();
		if ('Contacts' !== $sourceModule->getName()) {
			$sourceModule = $this->relationModel->getRelationModuleModel();
		}
		foreach ($this->customFields as $fieldName => $data) {
			$field = new \Vtiger_Field_Model();
			$field->set('name', $fieldName)->set('column', $fieldName)->set('table', static::TABLE_NAME)->set('fromOutsideList', true)->setModule($sourceModule);
			foreach ($data as $key => $value) {
				$field->set($key, $value);
			}
			$fields[$fieldName] = $field;
		}
		return $fields;
	}

	/**
	 * {@inheritdoc}
	 */
	public function getQuery()
	{
		parent::getQuery();
		$tableName = static::TABLE_NAME;
		$queryGenerator = $this->relationModel->getQueryGenerator();
		foreach (array_keys($this->customFields) as $fieldName) {
			$queryGenerator->setCustomColumn(["{$tableName}.{$fieldName}"]);
		}
	}

	/**
	 * {@inheritdoc}
	 */
	public function create(int $sourceRecordId, int $destinationRecordId): bool
	{
		$result = false;
		if (!$this->getRelationData($sourceRecordId, $destinationRecordId)) {
			$result = \App\Db::getInstance()->createCommand()->insert(static::TABLE_NAME, ['crmid' => $sourceRecordId, 'relcrmid' => $destinationRecordId])->execute();
		}

		return $result;
	}

	/**
	 * updateRelationData function.
	 *
	 * @param int   $sourceRecordId
	 * @param int   $destinationRecordId
	 * @param array $updateData
	 *
	 * @return bool
	 */
	public function updateRelationData(int $sourceRecordId, int $destinationRecordId, array $updateData): bool
	{
		$conditions = [
			'or',
			['crmid' => $sourceRecordId, 'relcrmid' => $destinationRecordId],
			['crmid' => $destinationRecordId, 'relcrmid' => $sourceRecordId]
		];
		$result = (bool) $this->getRelationData($sourceRecordId, $destinationRecordId);
		if ($result) {
			$result = (bool) \App\Db::getInstance()->createCommand()->update(static::TABLE_NAME, $updateData, $conditions)->execute();
		}
		return $result;
	}

	/**
	 * Get relation data.
	 *
	 * @param int $sourceRecordId
	 * @param int $destinationRecordId
	 *
	 * @return array
	 */
	public function getRelationData(int $sourceRecordId, int $destinationRecordId)
	{
		return (new \App\Db\Query())->from(static::TABLE_NAME)->where([
			'or',
			['crmid' => $sourceRecordId, 'relcrmid' => $destinationRecordId],
			['crmid' => $destinationRecordId, 'relcrmid' => $sourceRecordId]
		])->one();
	}
}

````

</details>

3 - Create relationship Contact <-> Contact using GUI and then manually change the relationship name from getRelatedList to GetRelatedMembers in vtiger_relatedlists (or edit if relationship already exists).

4 - Using phpmyadmin, you can then customize status_rel picklist values in vtiger_status_rel tab

That's it, it should work but I think changes made in vtiger_status_rel tab could be overwritten by a future YF upgrade.

Bonus : you can also [add this related tab as a widget](https://gitstable.yetiforce.com/index.php?module=Widgets&parent=Settings&view=Index)
